### PR TITLE
TWINE: new engine for Little Big Adventure 1

### PR DESCRIPTION
**Pull request by mgerhardy**
Ported Little Big Adventure twin-e which is no longer under active development (since 2015 afaik) to ScummVM.

The engine is still work-in-progress but I though that sharing it already might make sense to get feedback.

Midi Sound and music is ported. Input is using the `KeyMapper`. Rendering is using `Graphics::ManagedSurface`.

# TODO

- [ ] Refactor the input code - `struct Keyboard` - replace all those magic numbers and bitmask checks with proper constants
- [x] Fix the menu handling
- [ ] Port the Audio-CD code
- [ ] Replace `memset` with `Common::fill`
- [ ] Support autosaving and loading/saving via ui
- [x] Fix blitting (alpha issues for e.g. the character-  `TwinEEngine::copyBlockPhys`)

### CODE REVIEW

File: .clang-tidy [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
```
Comment (aquadran) [inline .clang-tidy:1]: non engine related changes must go as separate PR. also I saw open one already

Comment (mgerhardy) [reply .clang-tidy:1]: Yeah - will rebase this once i finish it. But for now I would like to have the `clang-tidy` warnings in my ide as well as the `create_project` patch (which is also a separate pr already). But leaving this here as a reminder

File: engines/twine/actor.cpp [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include <stdio.h>
```
Comment (anandmehrotra) [inline engines/twine/actor.cpp:23]: Any #include <..>  is not allowed for engines

Comment (dreammaster) [inline engines/twine/actor.cpp:23]: Any #include <..> lines are not allowed for engines.

File: engines/twine/actor.cpp [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+#include <stdlib.h>
+#include <string.h>
+
+#include "actor.h"
```
Comment (dreammaster) [inline engines/twine/actor.cpp:27]: Inclusion of engine files should also, by standard, include the engine's folder name, like in #include "twine/actor.h"

File: engines/twine/actor.cpp [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+	actor->info1 = 0;
+	actor->info2 = 0;
+	actor->info3 = 0;
+
+	actor->brickShape = 0;
+	actor->collision = -1;
+	actor->standOn = -1;
+	actor->zone = -1;
+
+	memset(&actor->staticFlags, 0, 2);
```
Comment (dreammaster) [inline engines/twine/actor.cpp:422]: It's recommended that engines use Common::fill rather than memset directly

File: engines/twine/actor.cpp [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+	actor->animType = 0;
+	actor->animPosition = 0;
+
+	_engine->_movements->setActorAngleSafe(0, 0, 0, &actor->move);
+
+	actor->positionInMoveScript = -1;
+	actor->positionInLifeScript = 0;
+}
+
+/** Process hit actor
```
Comment (dreammaster) [inline engines/twine/actor.cpp:443]: Method comments should be kept just in the header files rather than in the cpp files

File: engines/twine/animations.h [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	/** Process acotr animation actions
+	@param actorIdx Actor index */
+	void processAnimActions(int32 actorIdx);
+
+	/** Process main loop actor animations
+	@param actorIdx Actor index */
+	void processActorAnimations(int32 actorIdx);
+};
+
+}
```
Comment (dreammaster) [inline engines/twine/animations.h:156]: You're already using // namespace TwinE comments at the end of the namespaces elsewhere, so it's a good practice to have the comment for all of them

File: engines/twine/debug.cpp [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+			debugPlasmaWindow(text, color);
+
+			// quit
+			if (mouseData.right)
+				quit = 1;
+
+			fpsCycles(25); // rest
+
+			count++;
+		} while (!quit);
```
Comment (dreammaster) [inline engines/twine/debug.cpp:534]: Apart from maybe clarifying quit to be a bool rather than an int, something you'll need to do is hunt down every place where you have a loop like this and ensure that you break out of the loop if a call to the engine's shouldQuit() method returns true. Otherwise, the game will be prevented from exiting if the user tries to close the ScummVM window when execution is within the loop

File: engines/twine/actor.cpp (hunk 3 above)
Comment (mgerhardy) [reply engines/twine/actor.cpp:27]: thanks - good to know - will fix them.

File: engines/twine/debug.grid.cpp [hunk 8]
```diff
@@ -0,0 +1,126 @@
+/* ScummVM - Graphic Adventure Engine
```
Comment (dreammaster) [inline engines/twine/debug.grid.cpp:1]: I'm not comfortable with having a filename with multiple dots in it. Whilst I don't think there's any specific rule against it, it's more standard practice to use an underscore, like "debug_grid.cpp"

File: engines/twine/actor.cpp (hunk 2 above)
Comment (mgerhardy) [reply engines/twine/actor.cpp]: Resolving this after my latest commit - this is still the case in `fcaseopen.[cpp/h]` - but that is going to be removed anyway (by using the scummvm filesystem layer)

File: engines/twine/debug.grid.h [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+class TwinEEngine;
+
+class DebugGrid {
+private:
+	TwinEEngine *_engine;
+
+public:
+	DebugGrid(TwinEEngine *engine);
+
+	int32 useFreeCamera = 0;
```
Comment (dreammaster) [inline engines/twine/debug.grid.h:39]: Whilst my understanding is that C++11 allows field definitions like this, and we now support it for engines, you'll need to put the correct c++11 dependency requirement in the configure.engine

File: engines/twine/detection.cpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+/* ScummVM - Graphic Adventure Engine
```
Comment (dreammaster) [inline engines/twine/detection.cpp:23]: Double ScummVM comments

File: engines/twine/detection.h [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+/* ScummVM - Graphic Adventure Engine
```
Comment (dreammaster) [inline engines/twine/detection.h:23]: Doubled up comments here too

File: engines/twine/actor.cpp (hunk 5 above)
Comment (mgerhardy) [reply engines/twine/actor.cpp]: Should be fixed - thanks

File: engines/twine/animations.h (hunk 6 above)
Comment (mgerhardy) [reply engines/twine/animations.h]: woops - they were added by the formatter - not sure why it is missing here. will re-run the formatter

File: engines/twine/debug.grid.cpp (hunk 8 above)
Comment (mgerhardy) [reply engines/twine/debug.grid.cpp]: I've renamed them

File: engines/twine/detection.cpp (hunk 10 above)
Comment (mgerhardy) [reply engines/twine/detection.cpp]: Fixed

File: engines/twine/debug.grid.h (hunk 9 above)
Comment (mgerhardy) [reply engines/twine/debug.grid.h]: Added - thanks for the hint

File: engines/twine/debug.cpp (hunk 7 above)
Comment (mgerhardy) [reply engines/twine/debug.cpp:534]: True - but I think that I would rather like to convert these ui debug actions to console commands.

File: .clang-tidy (hunk 1 above)
Comment (mgerhardy) [reply .clang-tidy:1]: Added to the todo list at the pr description - so closing the reminder here.

File: engines/twine/fcaseopen.cpp [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "fcaseopen.h"
+
+#include <errno.h>
```
Comment (sev-) [inline engines/twine/fcaseopen.cpp:25]: These have to go.

In general, this whole file should be removed. Our File code is already case-insensitive.

File: engines/twine/fcaseopen.cpp [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "fcaseopen.h"
```
Comment (sev-) [inline engines/twine/fcaseopen.cpp:23]: `#include "twine/fcaseopen.h"`

File: engines/twine/filereader.cpp [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+ */
+
+#define FORBIDDEN_SYMBOL_ALLOW_ALL
+
+#include "twine/filereader.h"
+#include "twine/fcaseopen.h"
+
+namespace TwinE {
+
+bool frfeed(FileReader *fr) {
```
Comment (sev-) [inline engines/twine/filereader.cpp:30]: As I understand, you still plan to switch to our Stream class, so not commenting much here.

File: engines/twine/flamovies.cpp [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+
+	_fadeOut = -1;
+	fadeOutFrames = 0;
+
+	if (!fropen2(&frFla, (char *)fileNamePath, "rb"))
+		return;
+
+	workVideoBufferCopy = _engine->workVideoBuffer;
+
+	frread(&frFla, &flaHeaderData.version, 6);
```
Comment (sev-) [inline engines/twine/flamovies.cpp:277]: Yes, `Stream::readUint32LE()` should be used here for portability

File: engines/twine/gamestate.cpp [hunk 16]
```diff
@@ -0,0 +1,525 @@
...
+	namePtr = savePlayerName;
+
+	frread(&fr, &data, 1); // save game id
+
+	do {
+		frread(&fr, &data, 1); // get save player name characters
+		*(namePtr++) = data;
+	} while (data);
+
+	frread(&fr, &data, 1); // number of game flags, always 0xFF
```
Comment (sev-) [inline engines/twine/gamestate.cpp:190]: It would be best using `Common::Serializer` here.

File: engines/twine/interface.cpp [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+	else if (x > textWindowRight)
+		code |= RIGHT;
+	if (y < textWindowTop)
+		code |= TOP;
+	else if (y > textWindowBottom)
+		code |= BOTTOM;
+	return code;
+}
+
+void Interface::drawLine(int32 startWidth, int32 startHeight, int32 endWidth, int32 endHeight, int32 lineColor) {
```
Comment (sev-) [inline engines/twine/interface.cpp:50]: Will our `Graphics::drawLine()` from `graphics/primitives.h` be working here?

File: engines/twine/menu.cpp [hunk 18]
```diff
@@ -0,0 +1,1214 @@
...
+
+	_engine->_gameState->initEngineProjections();
+
+	_engine->cfgfile.LanguageCDId = tmpLanguageCD;
+
+	_engine->_text->initTextBank(_engine->_text->currentTextBank + 3);
+
+	while (_engine->_keyboard.skipIntro != 0 && _engine->_keyboard.skippedKey != 0) {
+		readKeys();
+		_engine->_system->delayMillis(1);
```
Comment (sev-) [inline engines/twine/menu.cpp:1210]: you potentially need also to call `g_system->updateScreen()`, so mouse gets redrawn.

File: engines/twine/movements.cpp [hunk 19]
```diff
@@ -0,0 +1,493 @@
...
+#include "twine/gamestate.h"
+#include "twine/grid.h"
+#include "twine/keyboard.h"
+#include "twine/renderer.h"
+#include "twine/scene.h"
+#include "twine/twine.h"
+
+namespace TwinE {
+// TODO: remove me
+#define PI 3.14159265
```
Comment (sev-) [inline engines/twine/movements.cpp:37]: we have `M_PI`

File: engines/twine/music.cpp [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include <SDL.h>
```
Comment (sev-) [inline engines/twine/music.cpp:23]: These need to go

File: engines/twine/redraw.cpp [hunk 21]
```diff
@@ -0,0 +1,802 @@
...
+#include "twine/renderer.h"
+#include "twine/resources.h"
+#include "twine/scene.h"
+#include "twine/screens.h"
+#include "twine/sdlengine.h"
+#include "twine/sound.h"
+#include "twine/text.h"
+
+#ifdef GAMEMOD
+#include "debug.scene.h"
```
Comment (sev-) [inline engines/twine/redraw.cpp:42]: `#include "twine/debug_scene.h"`

File: engines/twine/redraw.cpp [hunk 22]
```diff
@@ -0,0 +1,802 @@
...
+	}
+}
+
+void Redraw::sortDrawingList(DrawListStruct *list, int32 listSize) {
+	int32 i;
+	int32 j;
+
+	DrawListStruct tempStruct;
+
+	for (i = 0; i < listSize - 1; i++) {
```
Comment (sev-) [inline engines/twine/redraw.cpp:166]: I would write C++ `for (int i = 0;...`

File: engines/twine/shadeangletab.h [hunk 23]
```diff
@@ -0,0 +1,1059 @@
...
+
+#ifndef TWINE_SHADEANGLETAB_H
+#define TWINE_SHADEANGLETAB_H
+
+#include "common/scummsys.h"
+
+namespace TwinE {
+
+const int16 shadeAngleTable[1024] = {
+    0, // tab1
```
Comment (sev-) [inline engines/twine/shadeangletab.h:31]: If possible, I would reduce vertical size here, grouping the entries by 16 per line or so. But not a big deal.

File: engines/twine/sound.cpp [hunk 24]
```diff
@@ -0,0 +1,258 @@
...
+namespace TwinE {
+
+/** Samples chunk variable */
+Mix_Chunk *sample;
+
+void Sound::sampleVolume(int32 chan, int32 volume) {
+	Mix_Volume(chan, volume / 2);
+}
+
+void Sound::playFlaSample(int32 index, int32 frequency, int32 repeat, int32 x, int32 y) {
```
Comment (sev-) [inline engines/twine/sound.cpp:44]: Yes, I understand, still todo to move to Audio::Mixer. Not commenting.

File: engines/twine/twine.cpp [hunk 25]
```diff
@@ -0,0 +1,971 @@
...
+			if (!strcmp(LanguageTypes[i], buffer)) {
+				return i;
+			}
+		}
+	}
+
+	return 0; // English
+}
+
+void TwinEEngine::initConfigurations() {
```
Comment (sev-) [inline engines/twine/twine.cpp:210]: These will go to `ConfMan`.

File: engines/twine/xmidi.cpp [hunk 26]
```diff
@@ -0,0 +1,764 @@
...
+namespace TwinE {
+
+/**
+ * Provides comprehensive information on the next event in the MIDI stream.
+ * An EventInfo struct is instantiated by format-specific implementations
+ * of MidiParser::parseNextEvent() each time another event is needed.
+ *
+ * Adapted from the ScummVM project
+ */
+struct EventInfo {
```
Comment (sev-) [inline engines/twine/xmidi.cpp:36]: Please check, how well our `audio/midiparser_xmidi.cpp` is suitable for you.

**Review by sev- — COMMENTED** [review]
Good progress. If you need help or guidance with moving away from `filereader.cpp` to our `Common::Stream` code, or what to do with `sdlengine.cpp`, just say so.

File: engines/twine/movements.cpp (hunk 19 above)
Comment (mgerhardy) [reply engines/twine/movements.cpp:37]: I just wondered, because there are a lot of places where PI is hard coded in the engines. But I will just use M_PI then. Thanks

File: engines/twine/redraw.cpp (hunk 22 above)
Comment (mgerhardy) [reply engines/twine/redraw.cpp:166]: Yes, this was c89 code initially - will try to clean stuff up once I get there. There are a lot of other places left.

File: engines/twine/shadeangletab.h (hunk 23 above)
Comment (mgerhardy) [reply engines/twine/shadeangletab.h:31]: Right now I'm just applying the clang-format each time to the whole file. Once everything is in place and shape, I can do the final cleanup to match your style (or fix the clang-format file)

File: engines/twine/actor.h [hunk 27]
```diff
@@ -0,0 +1,304 @@
...
+
+/** Actors zone volumique points structure */
+typedef struct ZVPoint {
+	int16 bottomLeft;
+	int16 topRight;
+} ZVPoint;
+
+/** Actors zone volumique box structure */
+typedef struct ZVBox {
+	ZVPoint X;
```
[the remark this answers is not in the document]
Comment (mgerhardy) [reply engines/twine/actor.h]: Fixed.

File: engines/twine/animations.cpp [hunk 28]
```diff
@@ -0,0 +1,1228 @@
...
+				break;
+			}
+			case 2: { // dissallow global rotate + hide
+				applyAnimStep(&edi, eax, keyFrameLength);
+				applyAnimStep(&edi, eax, keyFrameLength);
+				applyAnimStep(&edi, eax, keyFrameLength);
+				break;
+			}
+			default: {
+				// TODO: printf("Unsupported animation rotation mode %d!\n", animOpcode);
```
[the remark this answers is not in the document]
Comment (mgerhardy) [reply engines/twine/animations.cpp]: Fixed

File: engines/twine/fcaseopen.cpp (hunk 13 above)
Comment (mgerhardy) [reply engines/twine/fcaseopen.cpp]: Fixed - because I've removed that file already

File: engines/access/events.cpp [hunk 29]
```diff
@@ -81,7 +81,7 @@ void EventsManager::setCursor(CursorType cursorId) {
 
 		// Create a surface to build up the cursor on
 		Graphics::Surface cursorSurface;
-		cursorSurface.create(16, 16, Graphics::PixelFormat::createFormatCLUT8());
+		cursorSurface.create(CURSOR_WIDTH, CURSOR_HEIGHT, Graphics::PixelFormat::createFormatCLUT8());
```
[the remark this answers is not in the document]
Comment (mgerhardy) [reply engines/access/events.cpp:84]: No, but I will still collect everything and rebase at the end to open separate pull requests. There are two other commits already, that should not be part of the final pull request

File: engines/twine/twine.cpp (hunk 25 above)
Comment (mgerhardy) [reply engines/twine/twine.cpp]: Done

File: engines/twine/sound.cpp (hunk 24 above)
Comment (mgerhardy) [reply engines/twine/sound.cpp:55]: Should be converted to `Audio::Mixer` now. I will see if it works once the lights are switched on for the first time.

**Comment by mgerhardy** [comment]
@sev- I would like to convert the `TwinEEngine::readKeys` method to use the `KeyMapper` from the start. But right now I don't have a clue on how to convert all these `SDL_PollEvent` calls to the `KeyMapper` actions. Is that the way to go? Or should I try to use `g_system->getEventManager()->pollEvent(event)` as a replacement for now?

**Comment by mgerhardy** [comment]
Palette handling is still broken.

Bildschirmfoto von 2020-10-20 13-13-03

I could need some help here.

```c++
void TwinEEngine::setPalette(uint8 *palette) {
    g_system->getPaletteManager()->setPalette(palette, 0, 256);
    flip();
}

void TwinEEngine::flip() {
    g_system->copyRectToScreen(frontVideoBuffer, SCREEN_WIDTH, 0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);
    g_system->updateScreen();
}
```

File: engines/access/events.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (mgerhardy) [reply engines/access/events.cpp]: This is in a separate PR now.

File: engines/twine/redraw.cpp (hunk 22 above)
Comment (mgerhardy) [reply engines/twine/redraw.cpp]: this is fixed now (but still not all locations)

**Comment by sev-** [comment]
Keymapper consists of 2 parts: registering keys and assigning them artificial ids. Then use normal `pollEvent()` and process those ids.

You may find the documentation here: https://wiki.scummvm.org/index.php?title=Keymapper

File: engines/twine/actor.h [hunk 30]
```diff
@@ -0,0 +1,342 @@
...
+enum HeroBehaviourType {
+	kNormal = 0,
+	kAthletic = 1,
+	kAggressive = 2,
+	kDiscrete = 3,
+	kProtoPack = 4
+};
+
+/** Actors move structure */
+typedef struct ActorMoveStruct {
```
Comment (sev-) [inline engines/twine/actor.h:55]: in C++ you may drop typedef and just write `struct ActionMoveStruct { ... };`

File: engines/twine/detection.h [hunk 31]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ */
+
+#ifndef TWINE_DETECTION_H
+#define TWINE_DETECTION_H
+
+namespace TwinE {
+
+enum GameFlag {
+	kGameFlagDemo = 1 << 0
```
Comment (sev-) [inline engines/twine/detection.h:29]: Could this be replaced with `ADGF_DEMO` constant which we already have defined?

**Review by sev- — COMMENTED** [review]
I looked for the second time till file menu.h

File: engines/twine/detection.h (hunk 31 above)
Comment (mgerhardy) [reply engines/twine/detection.h:29]: Fixed

File: engines/twine/actor.h (hunk 30 above)
Comment (mgerhardy) [reply engines/twine/actor.h]: Fixed

File: engines/twine/detection.h (hunk 31 above)
Comment (mgerhardy) [reply engines/twine/detection.h]: This is not used at all right now - I've removed it - was an artifact of the tucker engine that i've used as a template.

File: engines/twine/xmidi.cpp (hunk 26 above)
Comment (ccawley2011) [reply engines/twine/xmidi.cpp:37]: In addition, it looks like the game is using Miles, so you may want to take advantage of the Miles MIDI driver in `audio/miles.h` as well.

**Comment by sev-** [comment]
As discussed, merging this, so the work could continue in-tree.
